File: custom_consul_exporter/consul_metrics.py
import json
import re
import time
from urllib.request import urlopen

from prometheus_client import start_http_server
from prometheus_client.core import (
    REGISTRY,
    CounterMetricFamily,
    GaugeMetricFamily,
    SummaryMetricFamily,
)


def sanitize_name(s):
    return re.sub(r"[^a-zA-Z0-9:_]", "_", s)


class ConsulCollector(object):
    def collect(self):
        out = urlopen("http://localhost:8500/v1/agent/metrics").read()
        metrics = json.loads(out.decode("utf-8"))

        for g in metrics["Gauges"]:
            yield GaugeMetricFamily(
                sanitize_name(g["Name"]), "Consul metric " + g["Name"], g["Value"]
            )

        for c in metrics["Counters"]:
            yield CounterMetricFamily(
                sanitize_name(c["Name"]) + "_total",
                "Consul metric " + c["Name"],
                c["Count"],
            )

        for s in metrics["Samples"]:
            yield SummaryMetricFamily(
                sanitize_name(s["Name"]) + "_seconds",
                "Consul metric " + s["Name"],
                count_value=c["Count"],
                sum_value=s["Sum"] / 1000,
            )


if __name__ == "__main__":
    REGISTRY.register(ConsulCollector())
    start_http_server(8000)
    while True:
        time.sleep(1)

File: custom_consul_exporter/consul_metrics.go
package main

import (
	"github.com/hashicorp/consul/api"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"log"
	"net/http"
	"regexp"
)

var (
	up = prometheus.NewDesc(
		"consul_up",
		"Was talking to Consul successful.",
		nil, nil,
	)
	invalidChars = regexp.MustCompile("[^a-zA-Z0-9:_]")
)

type ConsulCollector struct{}

func (c ConsulCollector) Describe(ch chan<- *prometheus.Desc) {
	ch <- up // 发送 'up' 的指标描述
}

func (c ConsulCollector) Collect(ch chan<- prometheus.Metric) {
	consul, err := api.NewClient(api.DefaultConfig())
	if err != nil {
		ch <- prometheus.MustNewConstMetric(up, prometheus.GaugeValue, 0) // 向 Prometheus 发送失败的 'consul_up' 指标
	}

	metrics, err := consul.Agent().Metrics()
	if err != nil {
		ch <- prometheus.MustNewConstMetric(up, prometheus.GaugeValue, 0)
		return
	}
	ch <- prometheus.MustNewConstMetric(up, prometheus.GaugeValue, 1)

	for _, g := range metrics.Gauges {
		name := invalidChars.ReplaceAllLiteralString(g.Name, "_")
		desc := prometheus.NewDesc(name, "Consul metric "+g.Name, nil, g.Labels)
		ch <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, float64(g.Value))
	}

	for _, c := range metrics.Counters {
		name := invalidChars.ReplaceAllLiteralString(c.Name, "_")
		desc := prometheus.NewDesc(name+"_total", "Consul metric "+c.Name, nil, c.Labels)
		ch <- prometheus.MustNewConstMetric(desc, prometheus.CounterValue, float64(c.Count))
	}

	for _, s := range metrics.Samples {
		name := invalidChars.ReplaceAllLiteralString(s.Name, "_") + "_seconds"
		countDesc := prometheus.NewDesc(name+"_count", "Consul metric "+s.Name, nil, s.Labels)
		ch <- prometheus.MustNewConstMetric(countDesc, prometheus.CounterValue, float64(s.Count))
		sumDesc := prometheus.NewDesc(name+"_sum", "Consul metric "+s.Name, nil, s.Labels)
		ch <- prometheus.MustNewConstMetric(sumDesc, prometheus.CounterValue, s.Sum/1000)
	}
}

func main() {
	c := ConsulCollector{}
	prometheus.MustRegister(c)
	http.Handle("/metrics", promhttp.Handler())
	log.Fatal(http.ListenAndServe(":8000", nil))
}
